flask/cli.py

```diff
@@ -520,15 +520,37 @@ def load_dotenv():
 
     .. _python-dotenv: https://github.com/theskumar/python-dotenv#readme
 
+    :param names: List of file names to search for.
+    :param top_dir: Don't load files above this directory. If the current
+        directory is above this directory, it is set to this directory before
+        searching, then reset afterwards. This must be an absolute path.
+    :param chdir: Whether to set the current directory to the location of the
+        first file found.
+
     .. versionadded:: 1.0
     """
 
     if dotenv is None:
         return
 
+    old_dir = None
     new_dir = None
 
-    for name in ('.env', '.flaskenv'):
+    if top_dir is not None:
+        if not os.path.isabs(top_dir):
+            raise ValueError("'top_dir' must be an absolute path.")
+
+        cwd = os.path.realpath(os.getcwd())
+        top_dir = os.path.realpath(top_dir)
+
+        # If the top dir is below the current dir, store the current dir and
+        # change to the top dir temporarily. This disables setting the dir
+        # based on found env files.
+        if os.path.commonprefix((cwd, top_dir)) == cwd:
```

Comment:
This is backwards right now. What I intended was `common_prefix != top`, to find when the top dir is not an ancestor of the current dir. Also need to check if the prefix plus the rest of top dir is actually a director, like 3.6's `commonpath`.

When `top_dir` is not an ancestor, I don't want to permanently change it for the same reason mitsuhiko commented above: we don't know the relationship between the working dir and the top dir. For example, if the package is installed, the `.env` file might be in `/etc/myapp/config`, which probably isn't the right cwd either. Supervisor, uWSGI, and other runners have the ability to set the working dir if desired.

Basically, if the user specifies a non-ancestor top dir, I'm assuming they will be the ones managing the working directory.